### ISSUE TITLE
Segment fault/Bus error occurred (r:1600338617)

### ISSUE THREAD

**Issue by mohamed-barakat** [oscar-system/Singular.jl#270]
Executing the attached file with the Singular executable located in the artifacts directory causes the following segfault:

```
$ cd  ~/.julia/artifacts/f7d400612e5d95c1d35798c84cffe8e414a76823/bin/
$ ./Singular commands_file_of_Singular_with_PID_21298.txt
...
Singular : signal 8 (v: 4132):
current line:>>    q=q+facstd(qq[ll]);<<
Segment fault/Bus error occurred (r:1600338601)
please inform the authors
trying to restart...
Auf Wiedersehen.
$
```

This bug is related to https://github.com/oscar-system/GAP.jl/issues/519.

[commands_file_of_Singular_with_PID_21298.txt](https://github.com/oscar-system/Singular.jl/files/5238351/commands_file_of_Singular_with_PID_21298.txt)

**Comment by thofma** [comment]
I can reproduce it. It works with a vanilla Singular 4.1.3.

The Singular version that ships with Singular.jl
1) has NTL disabled. I think this means that some factoring functions are are not working.
2) is more recent then 4.1.3. This could also be the source of the problem, but I don't have a local Singular version which is up to date.

Could you have a look @tthsqe12? If you have a recent version of Singular checked out locally, I would first try to rule out 2).

I reproduced it as follows (just running the binaries as Mohamed did might not work because of problems with paths):
```julia
julia> using Singular_jll

julia> Singular_jll.Singular() do exe
         run(`$exe /tmp/commands_file_of_Singular_with_PID_21298.txt`) # `gdb --args $exe /tmp/commands_file_of_Singular_with_PID_21298.txt`
       end
```

**Comment by hannes14** [comment]
was a bug in the non-NTL factoring code,
fixed with https://github.com/Singular/Singular/commit/6f68e2baeb196eadcc0a546de4ad10c552f40b15

**Comment by mohamed-barakat** [comment]
Thank you very much @hannes14.

Is it possible to make a new release of `Singular_jll`?

**Comment by thofma** [comment]
Thanks @hannes14. I will do a new release tomorrow.

**Comment by mohamed-barakat** [comment]
Thanks, @thofma.

**Comment by thofma** [comment]
After a `]up` this should be working now.

**Comment by mohamed-barakat** [comment]
Thanks @thofma.

**Comment by thofma** [comment]
@mohamed-barakat Here is another way to work around the bug. I assume you have a working local Singular version lying around. Then you can use this in Singular.jl as follows:

Create a file in `$HOME/.julia/artifacts` called `Overrides.toml` with the following content:
```toml
[43d676ae-4934-50ba-8046-7a96366d613b]
libpolys = "/path/to/Singular"
libSingular = "/path/to/Singular"
libparse = "/path/to/Singular"
libfactory = "/path/to/Singular"
libsingular_resources = "/path/to/Singular"
libomalloc = "/path/to/Singular"
Singular = "/path/to/Singular"
```
(If you got the path wrong, he will complain about it when doing `using Singular` and show you where he has been looking for it. So it is easy to adjust the path until it works.)

**Comment by mohamed-barakat** [comment]
Thank you Tommy:

`HomalgProject.jl` sets the `PATH` and `LD_LIBRARY_PATH`/`DYLD_LIBRARY_PATH` to point the artifact paths of Singular and I have to disable this first. In any case doing what you suggest shows me the following warnings:

```julia
(@v1.5) pkg> build Singular
   Building LoadFlint → `~/.julia/packages/LoadFlint/10o48/deps/build.log`
   Building Nemo ─────→ `~/.julia/packages/Nemo/NVxdo/deps/build.log`
   Building Singular ─→ `~/.julia/packages/Singular/cuKCT/deps/build.log`

julia> using Singular
[ Info: Precompiling Singular [bcd08a7b-43d2-5ff7-b6d4-c458787f915c]
┌ Error: Invalid override in '/Users/mo/.julia/artifacts/Overrides.toml': Invalid SHA1 hash 'libpolys'
└ @ Pkg.Artifacts /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Pkg/src/Artifacts.jl:132
┌ Error: Invalid override in '/Users/mo/.julia/artifacts/Overrides.toml': Invalid SHA1 hash 'libsingular_resources'
└ @ Pkg.Artifacts /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Pkg/src/Artifacts.jl:132
┌ Error: Invalid override in '/Users/mo/.julia/artifacts/Overrides.toml': Invalid SHA1 hash 'libomalloc'
└ @ Pkg.Artifacts /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Pkg/src/Artifacts.jl:132
┌ Error: Invalid override in '/Users/mo/.julia/artifacts/Overrides.toml': Invalid SHA1 hash 'libSingular'
└ @ Pkg.Artifacts /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Pkg/src/Artifacts.jl:132
┌ Error: Invalid override in '/Users/mo/.julia/artifacts/Overrides.toml': Invalid SHA1 hash 'libparse'
└ @ Pkg.Artifacts /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Pkg/src/Artifacts.jl:132
┌ Error: Invalid override in '/Users/mo/.julia/artifacts/Overrides.toml': Invalid SHA1 hash 'libfactory'
└ @ Pkg.Artifacts /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Pkg/src/Artifacts.jl:132
┌ Error: Invalid override in '/Users/mo/.julia/artifacts/Overrides.toml': Invalid SHA1 hash 'Singular'
└ @ Pkg.Artifacts /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Pkg/src/Artifacts.jl:132
Singular.jl, based on
                     SINGULAR                                 /
 A Computer Algebra System for Polynomial Computations       /  Singular.jl: 0.4.1
                                                           0<   Singular   : 2.3.1-4
 by: W. Decker, G.-M. Greuel, G. Pfister, H. Schoenemann     \
FB Mathematik der Universitaet, D-67653 Kaiserslautern        \

(@v1.5) pkg>
```

**Comment by thofma** [comment]
I think I did it wrong (not sure why it still worked for me). I think the file needs to contain only:
```toml
[43d676ae-4934-50ba-8046-7a96366d613b]
Singular = "/home/blabla/test_install"
```

```julia
julia> using Singular_jll # do ]add Singular_jll for testing purposes

julia> Singular_jll.artifact_dir
"/home/blabla/test_install"

julia> Singular_jll.libsingular_path
"/home/blabla/test_install/lib/libSingular.so"
```
Could you try again?

**Comment by mohamed-barakat** [comment]
Thanks Tommy. If I do this I get the error

```
(@v1.5) pkg> build Singular
   Building LoadFlint → `~/.julia/packages/LoadFlint/10o48/deps/build.log`
   Building Nemo ─────→ `~/.julia/packages/Nemo/NVxdo/deps/build.log`
   Building Singular ─→ `~/.julia/packages/Singular/cuKCT/deps/build.log`
┌ Error: Error building `Singular`:
│ ERROR: LoadError: LoadError: IOError: could not spawn `~/software/spielwiese/current/bin/libparse ~/software/spielwiese/current/share/singular/LIB/GND.lib`: no such file or directory (ENOENT)
│ Stacktrace:
│  [1] _spawn_primitive(::String, ::Cmd, ::Array{Any,1}) at ./process.jl:99
│  [2] #585 at ./process.jl:112 [inlined]
│  [3] setup_stdios(::Base.var"#585#586"{Cmd}, ::Array{Any,1}) at ./process.jl:196
│  [4] _spawn at ./process.jl:111 [inlined]
│  [5] _spawn(::Base.CmdRedirect, ::Array{Any,1}) at ./process.jl:139 (repeats 2 times)
│  [6] run(::Base.CmdRedirect; wait::Bool) at ./process.jl:439
│  [7] run at ./process.jl:438 [inlined]
│  [8] (::var"#1#2"{String})(::String) at ~/.julia/packages/Singular/cuKCT/deps/parselibs.jl:7
│  [9] (::Singular_jll.var"#15#16"{var"#1#2"{String}})() at ~/.julia/packages/Singular_jll/i7rI4/src/wrappers/x86_64-apple-darwin14-cxx11.jl:99
│  [10] withenv(::Singular_jll.var"#15#16"{var"#1#2"{String}}, ::Pair{String,String}, ::Pair{String,String}) at ./env.jl:161
│  [11] libparse(::var"#1#2"{String}; adjust_PATH::Bool, adjust_LIBPATH::Bool) at ~/.julia/packages/Singular_jll/i7rI4/src/wrappers/x86_64-apple-darwin14-cxx11.jl:98
│  [12] libparse at ~/.julia/packages/Singular_jll/i7rI4/src/wrappers/x86_64-apple-darwin14-cxx11.jl:81 [inlined]
│  [13] execute at ~/.julia/packages/Singular/cuKCT/deps/parselibs.jl:2 [inlined]
│  [14] (::var"#6#9")(::IOStream) at ~/.julia/packages/Singular/cuKCT/deps/parselibs.jl:42
│  [15] open(::var"#6#9", ::String, ::Vararg{String,N} where N; kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at ./io.jl:325
│  [16] open(::Function, ::String, ::String) at ./io.jl:323
│  [17] (::var"#5#8")() at ~/.julia/packages/Singular/cuKCT/deps/parselibs.jl:38
│  [18] cd(::var"#5#8", ::String) at ./file.jl:104
│  [19] top-level scope at ~/.julia/packages/Singular/cuKCT/deps/parselibs.jl:37
│  [20] include(::String) at ./client.jl:457
│  [21] top-level scope at ~/.julia/packages/Singular/cuKCT/deps/build.jl:52
│  [22] include(::String) at ./client.jl:457
│  [23] top-level scope at none:5
│ in expression starting at ~/.julia/packages/Singular/cuKCT/deps/parselibs.jl:37
│ in expression starting at ~/.julia/packages/Singular/cuKCT/deps/build.jl:52
└ @ Pkg.Operations /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Pkg/src/Operations.jl:949
```

**Comment by thofma** [comment]
Ah sorry, I forgot to mention that you must have to build Singular with `--with-libparse`, i.e., run `./configure --with-libparse $yourotheroptions` and do `make install `again.

**Comment by mohamed-barakat** [comment]
This issue is now resolved with `Singular_jll v4.1.3+5`. I will close it.
